src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Private/JavaScriptMarshalImpl.ToJs.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace System.Runtime.InteropServices.JavaScript.Private
+{
+    internal partial class JavaScriptMarshalImpl
+    {
+        private unsafe delegate IntPtr MarshalToJsDelegate(object value, IntPtr stackRoot);
+        private static Dictionary<IntPtr, MarshalToJsDelegate> toJsMarshallers = new Dictionary<IntPtr, MarshalToJsDelegate>();
+
+#nullable disable
+        internal static unsafe IntPtr DynamicMarshalToJs<T>(IntPtr typeHandle, ref T value)
+        {
+            if (typeof(T).IsPrimitive)
+            {
+                // this is fast path for primitive value types, without boxing
+                return (IntPtr)Unsafe.AsPointer(ref value);
+            }
+            IntPtr stackRoot = (IntPtr)Unsafe.AsPointer(ref value);
+            //log.Call(null, $"DynamicMarshalToJs {(int)pValue}");
+            if (toJsMarshallers.TryGetValue(typeHandle, out var marshaller))
+            {
+                return marshaller.Invoke(value, stackRoot);
```

Comment:
This model seems to require the C# -> JS marshaling operation not allocate anything that has a lifetime, which seems questionable. If the marshaling is complex enough to require us to dynamically invoke something from a lookup table, is it really safe to assume it doesn't need anything like a temporary buffer?

Reply:
I will get there yet. I think marshalling structs may need dynamic buffer. Especially when marshalling from JS to C#.

Reply:
Added TODO to another place for now.